File: ios/SFCrime/Domain/Models/DispatchCall.swift
import Foundation
import CoreLocation

/// Represents a live 911 dispatch call from DataSF.
/// These are real-time calls with ~10-15 minute delay.
struct DispatchCall: Identifiable, Codable, Hashable {
    let id: Int
    let cadNumber: String
    let callTypeCode: String?
    let callTypeDescription: String?
    let priority: Priority?

    let receivedAt: Date
    let dispatchAt: Date?
    let onSceneAt: Date?
    let closedAt: Date?

    let coordinates: Coordinates?
    let locationText: String?
    let district: String?
    let disposition: String?

    /// Geographic coordinates for map display
    struct Coordinates: Codable, Hashable {
        let latitude: Double
        let longitude: Double

        var clLocationCoordinate: CLLocationCoordinate2D {
            CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
        }
    }

    enum CodingKeys: String, CodingKey {
        case id
        case cadNumber = "cad_number"
        case callTypeCode = "call_type_code"
        case callTypeDescription = "call_type_description"
        case priority
        case receivedAt = "received_at"
        case dispatchAt = "dispatch_at"
        case onSceneAt = "on_scene_at"
        case closedAt = "closed_at"
        case coordinates
        case locationText = "location_text"
        case district
        case disposition
    }

    /// Time since the call was received
    var timeAgo: String {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: receivedAt, relativeTo: Date())
    }

    /// Whether the call is still active (not closed)
    var isActive: Bool {
        closedAt == nil
    }
}

// MARK: - API Response

/// Paginated response for dispatch calls
struct DispatchCallsResponse: Codable {
    let calls: [DispatchCall]
    let nextCursor: String?
    let total: Int?

    enum CodingKeys: String, CodingKey {
        case calls
        case nextCursor = "next_cursor"
        case total
    }
}

File: backend/app/schemas/dispatch_call.py
"""Pydantic schemas for dispatch calls."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class Coordinates(BaseModel):
    """Geographic coordinates."""

    latitude: float
    longitude: float


class DispatchCallOut(BaseModel):
    """Dispatch call response schema."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    cad_number: str
    call_type_code: str | None = None
    call_type_description: str | None = None
    priority: str | None = None

    received_at: datetime
    dispatch_at: datetime | None = None
    on_scene_at: datetime | None = None
    closed_at: datetime | None = None

    coordinates: Coordinates | None = None
    location_text: str | None = None
    district: str | None = None
    disposition: str | None = None


class DispatchCallsResponse(BaseModel):
    """Paginated response for dispatch calls."""

    calls: list[DispatchCallOut]
    next_cursor: str | None = None
    total: int | None = None


class BoundingBox(BaseModel):
    """Bounding box for viewport queries."""

    min_lat: float = Field(..., ge=-90, le=90)
    min_lng: float = Field(..., ge=-180, le=180)
    max_lat: float = Field(..., ge=-90, le=90)
    max_lng: float = Field(..., ge=-180, le=180)
